File: apis_c#/AthleteApi/AthleteApi/Services/AttemptService.cs
using System.Data;
using Microsoft.Data.SqlClient;
using AthleteApi.Models;

namespace AthleteApi.Services
{
    // Implementación del servicio de intentos que maneja las operaciones CRUD para los intentos
    public class AttemptService : IAttemptService
    {
        private readonly string _connectionString;

        // Constructor que inicializa la cadena de conexión a la base de datos
        public AttemptService(IConfiguration configuration)
        {
            // Obtiene la cadena de conexión desde la configuración
            _connectionString = configuration?.GetConnectionString("DefaultConnection") ?? throw new ArgumentNullException(nameof(configuration), "Configuration cannot be null");
        }

        // Método para agregar un nuevo intento en la base de datos
        public async Task AddAttempt(Attempt attempt)
        {
            // Crea una nueva conexión SQL usando la cadena de conexión
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                // Crea un nuevo comando SQL para ejecutar el procedimiento almacenado 'sp_addattempt'
                using (SqlCommand cmd = new SqlCommand("sp_addattempt", conn))
                {
                    cmd.CommandType = CommandType.StoredProcedure;

                    // Agrega los parámetros necesarios para el procedimiento almacenado
                    cmd.Parameters.AddWithValue("@ParticipationId", attempt.ParticipationId);
                    cmd.Parameters.AddWithValue("@AttemptNumber", attempt.AttemptNumber);
                    cmd.Parameters.AddWithValue("@Type", attempt.Type);
                    cmd.Parameters.AddWithValue("@WeightLifted", attempt.WeightLifted);
                    cmd.Parameters.AddWithValue("@Success", attempt.Success);

                    // Abre la conexión a la base de datos
                    conn.Open();
                    // Ejecuta el comando de forma asíncrona
                    await cmd.ExecuteNonQueryAsync();
                }
            }
        }

        // Método para obtener una lista de intentos por torneo desde la base de datos con paginación
        public async Task<IEnumerable<Attempt>> GetAttemptsByTournament(int? tournamentId, string tournamentName, int pageNumber, int pageSize)
        {
            var attempts = new List<Attempt>();

            // Crea una nueva conexión SQL usando la cadena de conexión
            using (SqlConnection conn = new SqlConnection(_connectionString))
            {
                // Crea un nuevo comando SQL para ejecutar el procedimiento almacenado 'sp_getattemptsbytournament'
                using (SqlCommand cmd = new SqlCommand("sp_getattemptsbytournament", conn))
                {
                    cmd.CommandType = CommandType.StoredProcedure;

                    // Agrega los parámetros necesarios para el procedimiento almacenado
                    cmd.Parameters.AddWithValue("@TournamentId", tournamentId.HasValue ? (object)tournamentId.Value : DBNull.Value);
                    cmd.Parameters.AddWithValue("@TournamentName", tournamentName ?? (object)DBNull.Value);
                    cmd.Parameters.AddWithValue("@PageNumber", pageNumber);
                    cmd.Parameters.AddWithValue("@PageSize", pageSize);

                    // Abre la conexión a la base de datos
                    conn.Open();
                    // Ejecuta el comando y obtiene un lector de datos de forma asíncrona
                    using (SqlDataReader reader = await cmd.ExecuteReaderAsync())
                    {
                        // Lee los datos devueltos por el lector de datos
                        while (await reader.ReadAsync())
                        {
                            var attempt = new Attempt
                            {
                                Id = reader.GetInt32(0),
                                ParticipationId = reader.GetInt32(1),
                                AttemptNumber = reader.GetInt32(2),
                                Type = reader.GetString(3),
                                WeightLifted = reader.GetDouble(4),
                                Success = reader.GetBoolean(5) ? 1 : 0,
                                TournamentName = reader.GetString(6),
                                TournamentId = reader.GetInt32(7)
                            };
                            // Agrega el intento a la lista de intentos
                            attempts.Add(attempt);
                        }
                    }
                }
            }

            // Retorna la lista de intentos
            return attempts;
        }
    }
}

File: apis_py/app/services/attempt_service.py
from typing import List, Optional
from fastapi import Depends
from sqlalchemy.orm import Session
from sqlalchemy import text
from app.schemas.attempt import AttemptCreate, AttemptResponse
from app.models.attempt import Attempt
from app.database import get_db

class AttemptService:
    """
    Servicio para manejar operaciones relacionadas con intentos en la base de datos.
    
    Atributos:
        db (Session): Sesión de la base de datos para realizar las operaciones.
    """

    def __init__(self, db: Session = Depends(get_db)):
        """
        Inicializa el servicio con una sesión de la base de datos.
        
        Args:
            db (Session): Sesión de la base de datos inyectada automáticamente por FastAPI.
        """
        self.db = db

    def get_attempts_by_tournament(self, tournament_id: Optional[int], tournament_name: Optional[str], page_number: int, page_size: int) -> List[AttemptResponse]:
        """
        Obtiene una lista paginada de intentos por torneo desde la base de datos.

        Args:
            tournament_id (Optional[int]): Identificador del torneo.
            tournament_name (Optional[str]): Nombre del torneo.
            page_number (int): Número de la página a obtener.
            page_size (int): Cantidad de registros por página.

        Returns:
            List[AttemptResponse]: Lista de respuestas con la información de los intentos.
        """
        attempts = self.db.execute(
            text("EXEC sp_getattemptsbytournament @TournamentId = :tournament_id, @TournamentName = :tournament_name, @PageNumber = :page_number, @PageSize = :page_size"),
            {'tournament_id': tournament_id, 'tournament_name': tournament_name, 'page_number': page_number, 'page_size': page_size}
        ).fetchall()
        return [AttemptResponse.model_validate(attempt) for attempt in attempts]

    def add_attempt(self, attempt: AttemptCreate) -> AttemptResponse:
        """
        Agrega un nuevo intento en la base de datos.

        Args:
            attempt (AttemptCreate): Objeto que contiene los datos del intento a crear.

        Returns:
            AttemptResponse: Objeto que representa el intento creado.

        Raises:
            ValueError: Si el intento no cumple con las reglas de negocio.
        """
        self.db.execute(
            text("EXEC sp_addattempt @ParticipationId = :participation_id, @AttemptNumber = :attempt_number, @Type = :type, @WeightLifted = :weight_lifted, @Success = :success"),
            {'participation_id': attempt.participation_id, 'attempt_number': attempt.attempt_number, 'type': attempt.type, 'weight_lifted': attempt.weight_lifted, 'success': attempt.success}
        )
        self.db.commit()
        new_attempt = self.db.query(Attempt).filter_by(participation_id=attempt.participation_id, attempt_number=attempt.attempt_number, type=attempt.type).first()
        return AttemptResponse.model_validate(new_attempt)